cow_instrument/DetectorInfo.h

```diff
@@ -0,0 +1,190 @@
+#ifndef DETECTOR_INFO_H
+#define DETECTOR_INFO_H
+
+#include <memory>
+#include <vector>
+#include <stdexcept>
+#include <string>
+#include <sstream>
+#include <cmath>
+#include <boost/optional.hpp>
+#include "Detector.h"
+#include "IdType.h"
+#include "MaskFlags.h"
+#include "MonitorFlags.h"
+#include "L2s.h"
+#include "Command.h"
+#include "Spectrum.h"
+#include "cow_ptr.h"
+#include <mutex>
+#include <thread>
+
+namespace {
+
+template <typename U>
+void detectorRangeCheck(size_t detectorIndex, const U &container) {
+  if (detectorIndex >= container.size()) {
+    std::stringstream buffer;
+    buffer << "Detector at index " << detectorIndex << " is out of range";
+    throw std::out_of_range(buffer.str());
+  }
+}
+
+double distance(const V3D &a, const V3D &b) {
+  return std::sqrt(((a[0] - b[0]) * (a[0] - b[0])) +
+                   ((a[1] - b[1]) * (a[1] - b[1])) +
+                   ((a[2] - b[2]) * (a[2] - b[2])));
+}
+}
+
+template <typename InstTree> class DetectorInfo {
+public:
+  template <typename V>
+  explicit DetectorInfo(V &&instrumentTree)
+      : m_nDetectors(instrumentTree->nDetectors()),
+        m_isMasked(std::make_shared<MaskFlags>(m_nDetectors, Bool(false))),
+        m_isMonitor(std::make_shared<MonitorFlags>(m_nDetectors, Bool(false))),
+        m_l2(std::make_shared<L2s>(m_nDetectors)),
+        m_sourcePos(instrumentTree->sourcePos()),
+        m_samplePos(instrumentTree->samplePos()),
+        m_instrumentTree(instrumentTree) {
```

Comment:
Ok, does that imply that in general you do not intend to make `InstrumentTree` polymorphic (i.e., only for testing)?